Fixed ZZFeatureMap not accepting a list of entanglement

shravanpatel30 opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary
This pull request fixes #12432 

When a List[List(int)] or List[List[List[int]]] or List[List[List[List[int]]]] is passed to the entanglement in `ZZFeatureMap`, it returns an error. This should work like `TwoLocal` does. 

### Details and comments

Because both `ZFeatureMap` and `ZZFeatureMap` inherits from the `PauliFeatureMap` the problem was with how `PauliFeatureMap` handles the entanglement lists.  To generate proper entanglement layers `PauliFeatureMap` relies on the `get_entangler_map()` (the method which takes 3 arguments) method from `NLocal` which handles invalid inputs and generates properly formatted entanglement layers. But unlike `TwoLocal` where we specify rotation blocks and the entanglement blocks as separate arguments, for `PauliFeatureMap` the paulis are specified as a single list. Due to this, we need some additional logic to handle how entanglement layers are generated from the user specified entanglement list. 

To fix this I have added a helper method `_selective_entangler_map()` and also added `get_entangler_map()` method to `PauliFeatureMap`. The newly specified `get_entangler_map()` method will override the `get_entangler_map()` method from the `NLocal` class for List[List(int)] or List[List[List[int]]] or List[List[List[List[int]]]] but if any other input is passed to entanglement (like a str `full`) then we fall back on the method from `NLocal` to handle those cases.

With my fix all the below examples work as expected:

- If entanglement is a List[List(int)]:
```
N = 5
layer1 = (0,1)
layer2 = (1,0)

qc = PauliFeatureMap(N, reps=2, paulis=['Z', 'ZZ'], entanglement=[layer1, layer3], insert_barriers=True)
qc.decompose().draw('mpl')
```
<img width="939" alt="List List int" src="https://github.com/user-attachments/assets/9e9e58ad-7777-4dac-9ab8-96211286b68b">

-  If entanglement is a List[List[List[int]]] (This means that entanglement will change according to reps):
```
N = 5
layer1 = [(0,1,2)]
layer2 = [(3,2,1)]

qc = PauliFeatureMap(N, reps=3, paulis=['Z', 'ZZZ'], entanglement=[layer1, layer2], insert_barriers=True)
qc.decompose().draw('mpl', fold=-1)
```
<img width="940" alt="List List List int" src="https://github.com/user-attachments/assets/f2e60f67-ced4-47f5-9241-44b51c8a9e15">

- If entanglement is List[List[List[List[int]]]] (This means that entanglement will change according to reps and paulis):
```
N = 5
layers = [[[(0,1), (2,3)], [(1,2), (3,4)]], [[(1,0), (3,2)] , [(2,1), (4,3)]]]

qc = PauliFeatureMap(N, reps=3, paulis=['Z', 'ZZ', 'XX'], entanglement=layers, insert_barriers=True)
qc.decompose().draw('mpl')
```
<img width="940" alt="List List List List int" src="https://github.com/user-attachments/assets/12ccc706-5804-422e-999a-486f5fa41d90">

@nonhermitian please review this and let me know if I should make any further changes.

shravanpatel30 committed:
Added a get_entangler_map() and _selective_entangler_map() functions to PauliFeatureMap

shravanpatel30 committed:
Added some logic to handle the List[List[List[List[int]]]] case. Refer to the comments in the code.

Cryoris commented:
Thanks for the contribution @shravanpatel30! This is indeed a missing feature or wrong documentation. However, I'm not sure this is the right approach to solving it, as this selection might get very messy if you have Pauli connections with differing numbers of qubits. 

Is there a use-case for which you need specific entanglements? If yes, another approach could be to support a dictionary with `{num_qubits: entanglement}` pairs, e.g.
```python
entanglement = {
    1: [(0,), (2,)],
    2: [(0, 1), (1, 2)],
    3: [(0, 1, 2)]
}
zz_feature_map = ZZFeatureMap(.., paulis=["z", "zz", "zzz"], entanglement=entanglement)
```

shravanpatel30 commented:
Hi @Cryoris, the data encoding using a feature map is an active area of research and so being able to control the entanglement in a feature map is really important. This gives more flexibility in creating custom feature maps. 

Currently, classes like `TwoLocal`, `NLocal`, and others that handle encoding allow entanglement to be specified using various formats like List[List[int]], List[List[List[int]]], or List[List[List[List[int]]]]. If we switch to using a dictionary for `ZZFeatureMap` and `PauliFeatureMap`, could that introduce inconsistencies or issues, given that these classes inherit from `NLocal`? However, if you believe that using a dictionary wouldn’t cause any issues, I would be happy to implement this approach.

Also, I might be misunderstanding your concern, but I’m unclear on what you meant by:
> as this selection might get very messy if you have Pauli connections with differing numbers of qubits.

The current implementation works even when the entanglement is `layer1 = [(0,), (2,), (1,2), (0,1,2)]`:
```
from qiskit.circuit.library import PauliFeatureMap, ZZFeatureMap

N = 3
layer1 = [(0,), (2,), (1,2), (0,1,2)]

qc = PauliFeatureMap(N, reps=2, paulis=['Z', 'ZZ', 'ZZZ'], entanglement=[layer1], insert_barriers=True)
qc.decompose().draw('mpl')
```
<img width="935" alt="image" src="https://github.com/user-attachments/assets/559103fd-ee42-4a7a-99b6-eecb4024c1b2">

Cryoris commented:
> If we switch to using a dictionary for ZZFeatureMap and PauliFeatureMap, could that introduce inconsistencies or issues, given that these classes inherit from NLocal? However, if you believe that using a dictionary wouldn’t cause any issues, I would be happy to implement this approach.

It's good to think about this -- in this case I would say it is fine to have the subclass have a specific behavior, since it does have a special structure and the `list[list[list...` does not apply here.

> Also, I might be misunderstanding your concern, but I’m unclear on what you meant by:
>> as this selection might get very messy if you have Pauli connections with differing numbers of qubits.

> The current implementation works even when the entanglement is layer1 = [(0,), (2,), (1,2), (0,1,2)]

I mean that the entanglement list can (1) get very large if we have all layers in one list, and (2) it can get difficult to read as one could even accidentially mix the layers, such as `[(0,1), (1,2), (2,3), (1,2,3), (3,4), (0,1,2), (0,),...]`.

shravanpatel30 committed:
`entanglement` can now be specified as dictionary `{num_qubits: entanglement}`

shravanpatel30 committed:
Added logic to handle cases when entanglement is not specified for some pauli blocks.

shravanpatel30 commented:
Hi @Cryoris,

I have made some changes to the `PauliFeatureMap` so that it now supports a dictionary `{num_qubits: entanglement}`.

For `entanglement = {
    1: [(0,), (2,)],
    2: [(0, 1), (1, 2)],
    3: [(0, 1, 2)]
}`:
<img width="935" alt="image" src="https://github.com/user-attachments/assets/9be7ecb8-bd80-4912-8e81-a3c3ad733a77">

And if the  `entanglement = {
    1: [(0,), (2,)],
    3: [(0, 1, 2)]
}`, then `full` entanglement (which is default for `PauliFeatureMap`) is used for the 2-qubit pauli `ZZ`:
<img width="935" alt="image" src="https://github.com/user-attachments/assets/3794ebc3-8bb3-45a8-b851-990cfef7667a">

If you are fine with these changes then I'll go ahead and modify the typing hints and docstrings of `PauliFeatureMap`, `ZFeatureMap` and `ZZFeatureMap`. I'll also add few tests and releasenotes for these changes.

Cryoris commented:
Thanks for the update! Could you update the code such that there's no default `full` entanglement when the number of qubits is not specified? I think it would be better if the code is as explicit as possible, instead of adding entanglement that is not specified. Users can always add the full entanglement if it's needed 🙂 

And yes to tests and releasenotes 👍🏻  The section `features_circuits` should be the right section.

shravanpatel30 committed:
Added tests cases for `PauliFeatureMap`

shravanpatel30 committed:
Added test and releasenotes. Also modified the typing hints and doc strings to reflect new changes.

shravanpatel30 commented:
Hi @Cryoris, I have added tests and releasenotes. I have also updated the typing hints and docstrings to reflect new changes. Let me know if you want me to make any further changes.

shravanpatel30 committed:
Replaced all tab characters with spaces to fix the linting error.

shravanpatel30 commented:
Sorry about the failing test.

I replaced all the tab characters with spaces as suggested by the linting error, but it's still failing. When I check the compiled release notes on my PC, my recently added release note appears correctly, so I'm not sure where the problem lies.

ElePT commented on releasenotes/notes/paulifeaturemap-takes-dictionary-as-entanglement-02037cb2d46e1c41.yaml:
```diff
@@ -0,0 +1,21 @@
+            qc.decompose().draw('mpl')
+
+
+
```
```suggestion
---
features_circuits:
  - |
    The entanglement argument in :class:`.PauliFeatureMap` and :class:`.ZZFeatureMap` will no longer support ``List[int]`` or ``List[List[int]]``. Instead,
    the entanglement structure in :class:`.PauliFeatureMap` and :class:`.ZZFeatureMap`  can now be specified as a dictionary where the keys represent  
    the number of qubits, and the values are lists of integer tuples that define which qubits are entangled with one another. This
    allows for more flexibility in constructing feature maps tailored to specific quantum algorithms. 
    Example usage::

      from qiskit.circuit.library import PauliFeatureMap
      entanglement = {
        1: [(0,), (2,)],
        2: [(0, 1), (1, 2)],
        3: [(0, 1, 2)],
      }
      params=[0, 1, 3.14]
      qc = PauliFeatureMap(3, reps=2, paulis=['Z', 'ZZ', 'ZZZ'], entanglement=entanglement, insert_barriers=True)
      qc.decompose().draw('mpl')
```

ElePT reviewed (commented):
Hi @shravanpatel30, I have left a suggestion that might fix the issue in the reno, I think the cause of the lint complaint is related to the indent in the code example. I have also added a few formatting suggestions for the referenced class names. You can directly click on "commit suggestion" to try it out.

shravanpatel30 committed:
Formatted releasenotes as per the suggestion to fix the issue in reno.

shravanpatel30 commented:
Hi @ElePT, thanks for the suggestion. Both code and the release notes now pass all the checks. Let me know if I need to make any more changes.

Cryoris commented on qiskit/circuit/library/data_preparation/pauli_feature_map.py:
```diff
@@ -116,7 +115,7 @@ def __init__(
         feature_dimension: Optional[int] = None,
         reps: int = 2,
-        entanglement: Union[str, List[List[int]], Callable[[int], List[int]]] = "full",
+        entanglement: Union[str, Dict[int, List[Tuple[int]]], Dict[int, List[List[int]]]] = "full",
```
This should still accept a callable, which can return an entanglement specific for a repetition. So the signature should be: `Union[str, Dict[...], Callable[[int], Union[str | Dict[...]]`.

Cryoris commented on qiskit/circuit/library/data_preparation/pauli_feature_map.py:
```diff
@@ -156,6 +158,7 @@ def __init__(
         self._data_map_func = data_map_func or self_product
         self._paulis = paulis or ["Z", "ZZ"]
         self._alpha = alpha
+        self.entanglement = entanglement
```
Is this necessary? This attribute should already be accessible via `self._entanglement`.

Cryoris commented on qiskit/circuit/library/data_preparation/pauli_feature_map.py:
```diff
@@ -281,6 +284,40 @@ def cx_chain(circuit, inverse=False):
+                                f"For num_qubits = {qb}, entanglement must be a "
+                                f"tuple of length {qb}. You specified {en}."
+                            )
+                        self.entanglement[qb][ind] = tuple(map(int, en))
```
I'm not sure we need this; users should already provide integers (as per the type hint) and I think the code works with both tuples and lists.

Cryoris reviewed (commented):
Looks almost good, thanks for the adjustments!

shravanpatel30 committed:
Added `Callable[[int], Union[str | Dict[...]]]` to the typing hints and included some logic to handle that.

shravanpatel30 commented:
@Cryoris, I have updated the typing hints and added some code to handle entanglement specified as a callable. I have also incorporated other changes in the code according to your comments.

Cryoris commented on qiskit/circuit/library/data_preparation/pauli_feature_map.py:
```diff
@@ -116,7 +115,12 @@ def __init__(
+            str,
+            Dict[int, List[Tuple[int]]],
+            Dict[int, List[List[int]]],
+            Callable[[int], Union[str, Dict[int, List[Tuple[int]]], Dict[int, List[List[int]]]]],
```
We only need one of `List[List]` or `List[Tuple]`, they are equivalent 🙂  (same for the ZZ feature map)
```suggestion
            Dict[int, List[Tuple[int]]],
            Callable[[int], Union[str, Dict[int, List[Tuple[int]]]]],
```

Cryoris commented on releasenotes/notes/paulifeaturemap-takes-dictionary-as-entanglement-02037cb2d46e1c41.yaml:
```diff
@@ -0,0 +1,20 @@
+      }
+      qc = PauliFeatureMap(3, reps=2, paulis=['Z', 'ZZ', 'ZZZ'], entanglement=entanglement, insert_barriers=True)
+      qc.decompose().draw('mpl')
+
```
It was a bug that we could specify the entanglement as lists to begin with -- that didn't make sense 🙂 I reformulated the reno a bit to reflect this:
```suggestion
fixes:
  - |
    Fixed that the entanglement in :class:`.PauliFeatureMap` and :class:`.ZZFeatureMap`
    could be given as ``List[int]`` or ``List[List[int]]``, which was incompatible with the fact
    that entanglement blocks of different sizes are used. Instead, the entanglement can be 
    given as dictionary with ``{block_size: entanglement}`` pairs.
features_circuits:
  - |
    :class:`.PauliFeatureMap` and :class:`.ZZFeatureMap` now support specifying the 
    entanglement as a dictionary where the keys represent the number of qubits, and 
    the values are lists of integer tuples that define which qubits are entangled with one another. This
    allows for more flexibility in constructing feature maps tailored to specific quantum algorithms. 
    Example usage::

      from qiskit.circuit.library import PauliFeatureMap
      entanglement = {
        1: [(0,), (2,)],
        2: [(0, 1), (1, 2)],
        3: [(0, 1, 2)],
      }
      qc = PauliFeatureMap(3, reps=2, paulis=['Z', 'ZZ', 'ZZZ'], entanglement=entanglement, insert_barriers=True)
      qc.decompose().draw('mpl')

```

shravanpatel30 committed:
Removed `List[List[int]]` from typing hints as this is equivalent to List[Tuple[int]]

Co-authored-by: Julien Gacon <gaconju@gmail.com>

shravanpatel30 committed:
updated the release notes

Co-authored-by: Julien Gacon <gaconju@gmail.com>

shravanpatel30 committed:
Updated typing hints in `ZZFeatureMap`

shravanpatel30 commented:
@Cryoris, I have made all the requested changes.

Cryoris reviewed (approved):
Thanks for the fix & new feature!
